### PR TITLE
Dra 400 check filter for preservica7

### PR DESCRIPTION
**Pull request by VictorHarbo**
Refactoring of `OaiResponseFilter`. The old implementation didn't really make use of the way we're supporting multiple filters. I've renamed the old `OaiResponseFilterPreservica` to `OaiResponseFilterPreservicaFive` and then created a new filter where the logic for Preservica 7 has been moved to.

As we are moving towards a preservica 7 world, I've chosen to let the filter in YAML files specified as: `filter: preservica` reference the newly added `OaiResponseFilterPreservicaSeven` class and introduced a `filter: preservica5` for backwards compatibility. 

While testing this filtering locally on jetty, with a local ds-storage. I've stumpled upon an error when harvesting from preservica. I get an `HTTP 401`-error with the comment: _WWW-Authenticate header missing for response code 401_. However I've managed to create an integration test, that passes this. Could you please update your aegis configuration, start ds-storage and try to run the integrationtest: `dk.kb.datahandler.oai.OaiHarvestClientIntegrationTest#testPreservicaSevenAuth` and afterwards confirm that a call through the OpenAPI interface returns a 401 error? (Btw, this error also occurs on the main branch, when I'm running locally.)

### CODE REVIEW

File: src/main/java/dk/kb/datahandler/oai/OaiHarvestClient.java [hunk 1]
```diff
@@ -65,13 +68,14 @@ public OaiResponse next() throws Exception{
         String metadataPrefix= oaiTarget.getMetadataprefix();
 
         uri=addQueryParamsToUri(uri, set, resumptionToken,metadataPrefix,from, until);
-         log.info("calling uri:"+uri);
+        log.info("calling uri:"+uri);
         //log.info("resumption token at:"+resumptionToken);
-        String xmlResponse=getHttpResponse(uri,oaiTarget.getUsername(),oaiTarget.getPassword()); 
+        String xmlResponse = getHttpResponse(uri, oaiTarget.getUsername(), oaiTarget.getPassword());
+        log.info("Does the test get here");
```
Comment (tokee) [inline src/main/java/dk/kb/datahandler/oai/OaiHarvestClient.java:74]: This does not look like production logging :grin:

File: src/main/java/dk/kb/datahandler/oai/OaiHarvestClient.java [hunk 2]
```diff
@@ -84,9 +88,9 @@ public OaiResponse next() throws Exception{
         String  resumptionToken=  getResumptionToken(document);
         oaiResponse.setTotalRecords(getResumptionTotalSize(document));        
 
-        if (resumptionToken != null && !resumptionToken.equals("")) {
+        if (resumptionToken != null && !resumptionToken.isEmpty()) {
             this.resumptionToken = resumptionToken;  
-      log.info("next resumption token:"+resumptionToken);
+            log.info("next resumption token:"+resumptionToken);
```
Comment (tokee) [inline src/main/java/dk/kb/datahandler/oai/OaiHarvestClient.java:93]: Not your code, but could you lower this to `debug`? Seems much too spammy for my taste and the resumption token is logged indirectly at line 71.

File: src/main/java/dk/kb/datahandler/oai/OaiHarvestClient.java [hunk 3]
```diff
@@ -231,11 +233,21 @@ protected PasswordAuthentication getPasswordAuthentication() {
...
+        /* The Preservica 7 OAI-PMH endpoint is acting up. Per their documentation they say that they follow the standard
+            and makes Basic authorixation possible. However, the response does not contain a WWW-Authenticate header, when queried.
+
+            If using their acces endpoint to get an access-token then the service works. I got an access token by calling the following:
+            curl -X 'POST'   'DEVELSERVER'   -H 'accept: application/json'   -H 'Content-Type: application/x-www-form-urlencoded'   -d 'username=USERNAME&password=PASSWORD&cookie=false&includeUserDetails=false'
+            Changing DEVELSERVER, USERNAME and PASSWORD
 
+            I've written the integration test dk.kb.datahandler.oai.OaiHarvestClientIntegrationTest.testPreservicaSevenAuth
+            which makes it seem like this might be a problem in our end as the test parses.
+         */
```
Comment (tokee) [inline src/main/java/dk/kb/datahandler/oai/OaiHarvestClient.java:245]: Is this still relevant after our pair-review test?

File: src/main/java/dk/kb/datahandler/oai/OaiHarvestClient.java [hunk 4]
```diff
@@ -231,11 +233,21 @@ protected PasswordAuthentication getPasswordAuthentication() {
...
         HttpRequest request = HttpRequest.newBuilder()                          
-                .uri(URI.create(uri))              
-                .header("User-Agent", "Java 11 HttpClient Bot")            
-                .header("Authorization", getBasicAuthenticationHeader(user, password))                
+                .uri(URI.create(uri))
+                //.header("Preservica-Access-Token", "b8844076-d96d-4fe1-b8fe-e0dd6bee5bbc")
```
Comment (tokee) [inline src/main/java/dk/kb/datahandler/oai/OaiHarvestClient.java:248]: This does not look like a thing that should be in the code.

File: src/main/java/dk/kb/datahandler/oai/OaiResponseFilterPreserviceSeven.java [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Filtering and delivery of Preservica OAI records from Preservica 7. Generates {@code datasource} prefixed IDs,
+ * derives type from {@link DsRecordDto#getId()} and resolves {@code parent} from content.
+ */
+public class OaiResponseFilterPreserviceSeven extends OaiResponseFilter{
```
Comment (tokee) [inline src/main/java/dk/kb/datahandler/oai/OaiResponseFilterPreserviceSeven.java:17]: Preservic**a**, not Preservic**e**.

File: src/main/java/dk/kb/datahandler/oai/OaiResponseFilterPreserviceSeven.java [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        } else {
+            log.warn("No specific origin has been extracted for preservica record '{}'.",
+                    oaiRecord.getId());
+            // Not quite sure what we should do in the case where nothing gets matched.
+            return datasource;
+        }
+    }
+
+    @Override
+    public String getParentID(OaiRecord oaiRecord, String origin) {
```
Comment (tokee) [inline src/main/java/dk/kb/datahandler/oai/OaiResponseFilterPreserviceSeven.java:122]: Is this still relevant for Preservica 7?

File: src/main/openapi/ds-datahandler-openapi_v1.yaml [hunk 7]
```diff
@@ -330,7 +330,7 @@ components:
           example: 'H.C. Andersen'
         filter:
           type: string
-          enum:  ['direct', 'preservica']
+          enum:  ['direct', 'preservica', 'preservica5']
```
Comment (tokee) [inline src/main/openapi/ds-datahandler-openapi_v1.yaml:333]: Please add a description that explain what `preservica` is, compared to `preservica5`.

File: src/test/java/dk/kb/datahandler/oai/OaiHarvestClientIntegrationTest.java [hunk 8]
```diff
@@ -59,4 +83,42 @@ void testHarvest1000OaiRecords() throws Exception {
...
+
+        YAML conf = ServiceConfig.getConfig();
+
+        // name: pvica6.devel
+        OaiTargetDto oaiTarget = new OaiTargetDto();
+        oaiTarget.setUrl(conf.getString("integration.oaiTargets[1].url")); //Public KB service
+        oaiTarget.setMetadataprefix(conf.getString("integration.oaiTargets[1].metadataPrefix"));
+        oaiTarget.setUsername(conf.getString("integration.oaiTargets[1].user"));
+        oaiTarget.setPassword(conf.getString("integration.oaiTargets[1].password"));;
+        oaiTarget.setDatasource(conf.getString("integration.oaiTargets[1].datasource"));
```
Comment (tokee) [inline src/test/java/dk/kb/datahandler/oai/OaiHarvestClientIntegrationTest.java:99]: Could this be written with a conditional yPath instead? `[name=pvica7.devel]`?

**Review by tokee — CHANGES_REQUESTED** [review]
Tiny suggestions and a little "Remember the changelog".

File: src/main/java/dk/kb/datahandler/oai/OaiHarvestClient.java (hunk 3 above)
Comment (VictorHarbo) [reply src/main/java/dk/kb/datahandler/oai/OaiHarvestClient.java:245]: Nope! Thanks!

File: src/main/java/dk/kb/datahandler/oai/OaiHarvestClient.java (hunk 4 above)
Comment (VictorHarbo) [reply src/main/java/dk/kb/datahandler/oai/OaiHarvestClient.java:248]: Correct, I used the preservica access api to test, while I couldn't run the service. Somehow the access token could be used over VPN while basic auth couldn't

File: src/main/java/dk/kb/datahandler/oai/OaiResponseFilterPreserviceSeven.java (hunk 5 above)
Comment (VictorHarbo) [reply src/main/java/dk/kb/datahandler/oai/OaiResponseFilterPreserviceSeven.java:17]: Nice catch

File: src/main/openapi/ds-datahandler-openapi_v1.yaml (hunk 7 above)
Comment (tokee) [reply src/main/openapi/ds-datahandler-openapi_v1.yaml:333]: Very nice now!

File: src/test/java/dk/kb/datahandler/oai/OaiHarvestClientIntegrationTest.java (hunk 8 above)
Comment (tokee) [reply src/test/java/dk/kb/datahandler/oai/OaiHarvestClientIntegrationTest.java:99]: Yeah... Looks like an issue for `kb-util`. I have created https://github.com/kb-dk/kb-util/issues/67

**Pull request merged**